### PR TITLE
avifgainmaputil: Fix MSVC type conversion warnings

### PR DESCRIPTION
**Pull request by wantehchang**
Fix MSVC C4244 warnings on type conversions, possible loss of data.

### CODE REVIEW

**Comment by wantehchang** [comment]
Here are the MSVC warning messages:

```
D:\a\libavif\libavif\apps\avifgainmaputil\combine_command.cc(102): warning C4244: 'argument': conversion from 'float' to 'const _Ty', possible loss of data
        with
        [
            _Ty=uint32_t
        ]
D:\a\libavif\libavif\apps\avifgainmaputil\combine_command.cc(104): warning C4244: 'argument': conversion from 'float' to 'const _Ty', possible loss of data
        with
        [
            _Ty=uint32_t
        ]
```

File: apps/avifgainmaputil/tonemap_command.cc [hunk 1]
```diff
@@ -52,7 +52,7 @@ avifResult TonemapCommand::Run() {
   avifContentLightLevelInformationBox clli_box = {};
   bool clli_set = false;
   if (arg_clli_str_.value().size() > 0) {
-    std::vector<uint32_t> clli;
+    std::vector<uint16_t> clli;
```
Comment (wantehchang) [inline apps/avifgainmaputil/tonemap_command.cc:55]: At lines 62-63 these are assigned to `clli_box.maxCLL` and `clli_box.maxPALL`, both of which are `uint16_t`.

So declaring the `clli` vector as `uint16_t` fixed the warnings.

If the `uint16_t` type is too small, we can change lines 62-63 to:

```
    if (clli[0] > UINT16_MAX || clli[1] > UINT16_MAX) {
      std::cerr << "Huge clli values > 2^16 - 1\n";
      return AVIF_RESULT_INVALID_ARGUMENT;
    }
    clli_box.maxCLL = static_cast<uint16_t>(clli[0]);
    clli_box.maxPALL = static_cast<uint16_t>(clli[1]);
```

Comment (vigneshvg) [reply apps/avifgainmaputil/tonemap_command.cc:55]: i think this the current cast is fine.

**Review by vigneshvg — APPROVED** [review]

File: apps/avifgainmaputil/combine_command.cc [hunk 2]
```diff
@@ -98,10 +98,13 @@ avifResult CombineCommand::Run() {
...
-      std::round((float)base_image->width / downscaling), 1u);
-  const uint32_t gain_map_height = std::max<uint32_t>(
-      std::round((float)base_image->height / downscaling), 1u);
+  const uint32_t gain_map_width = std::max(
+      static_cast<uint32_t>(std::round((float)base_image->width / downscaling)),
+      1u);
+  const uint32_t gain_map_height =
+      std::max(static_cast<uint32_t>(
+                   std::round((float)base_image->height / downscaling)),
+               1u);
```
[the remark this answers is not in the document]
Comment (wantehchang) [reply apps/avifgainmaputil/combine_command.cc]: With the casts, this code is quite verbose. I just realized that we can do the rounded division using integer arithmetic. Please see the second commit in this pull request.

File: apps/avifgainmaputil/combine_command.cc [hunk 3]
```diff
@@ -98,13 +98,11 @@ avifResult CombineCommand::Run() {
   }
 
   const int downscaling = std::max<int>(1, arg_downscaling_);
-  const uint32_t gain_map_width = std::max(
-      static_cast<uint32_t>(std::round((float)base_image->width / downscaling)),
-      1u);
+  const int rounding = downscaling / 2;
+  const uint32_t gain_map_width =
+      std::max((base_image->width + rounding) / downscaling, 1u);
```
Comment (y-guyon) [inline apps/avifgainmaputil/combine_command.cc:103]: `width` is `uint32_t`, `rounding` and `downscaling` should be unsigned too, no?

**Review by y-guyon — APPROVED** [review]

File: apps/avifgainmaputil/combine_command.cc (hunk 3 above)
Comment (wantehchang) [reply apps/avifgainmaputil/combine_command.cc:103]: Done. When `int` and `unsigned int` are used in the same operation, `int` is implicitly converted to `unsigned int`. It's good to use the desired type explicitly.

Comment (y-guyon) [reply apps/avifgainmaputil/combine_command.cc:103]: I remember compiler warnings on Windows complaining about signed/unsigned mismatch. I understand it is fine according to the C standard though.

Comment (wantehchang) [reply apps/avifgainmaputil/combine_command.cc:103]: Compilers warn about signed/unsigned mismatch in comparisons. I have not seem compiler warnings about signed/unsigned mismatch in arithmetic operations (`+`, `-`, `*`, `/`).

**Pull request merged**